web/apps/labelstudio/src/app/App.jsx

```diff
@@ -17,18 +18,30 @@ import "./App.styl";
 import { AsyncPage } from "./AsyncPage/AsyncPage";
 import ErrorBoundary from "./ErrorBoundary";
 import { RootPage } from "./RootPage";
-import { FF_OPTIC_2, isFF } from "../utils/feature-flags";
+import { FF_OPTIC_2, FF_UNSAVED_CHANGES, isFF } from "../utils/feature-flags";
 import { ToastProvider, ToastViewport } from "../components/Toast/Toast";
 
 const baseURL = new URL(APP_SETTINGS.hostname || location.origin);
+export const UNBLOCK_HISTORY_MESSAGE = "UNBLOCK_HISTORY";
 
 const browserHistory = createBrowserHistory({
   basename: baseURL.pathname || "/",
   getUserConfirmation: (message, callback) => {
+    // `history.block` doesn't block events, so in the case of listeners,
+    // we need to have some flag that can be checked for preventing related actions
+    // `isBlocking` flag is used for this purpose
+    browserHistory.isBlocking = true;
+    const callbackWrapper = (result) => {
+      browserHistory.isBlocking = false;
+      callback(result);
+      isFF(FF_UNSAVED_CHANGES) && window.postMessage({ source: "label-studio", payload: UNBLOCK_HISTORY_MESSAGE });
+    };
     if (isFF(FF_OPTIC_2) && message === DRAFT_GUARD_KEY) {
-      draftGuardCallback.current = callback;
+      draftGuardCallback.current = callbackWrapper;
+    } else if (isFF(FF_UNSAVED_CHANGES) && message === LEAVE_BLOCKER_KEY) {
+      leaveBlockerCallback.current = callbackWrapper;
```

Comment:
can we somehow use only one logic for both cases? it's getting messy...

Reply:
This would require refactoring the mechanisms used for blocking/restoration of the DraftGuard, to use the LeaveBlocker instead. Seems like that would be a great option, not sure the amount of changes required would be something tenable inside of the same PR.